libsolidity/analysis/PostTypeChecker.cpp

```diff
@@ -422,6 +433,60 @@ struct ReservedErrorSelector: public PostTypeChecker::Checker
 	}
 };
 
+class LValueChecker: public ASTConstVisitor
+{
+public:
+	LValueChecker(Identifier const& _identifier): m_declaration(_identifier.annotation().referencedDeclaration) {}
+	bool willBeWrittenTo() const { return m_willBeWrittenTo; }
+	void endVisit(Identifier const& _identifier) override
+	{
+		if (
+			_identifier.annotation().referencedDeclaration == m_declaration &&
+			_identifier.annotation().willBeWrittenTo
+		)
+			m_willBeWrittenTo = true;
+	}
+private:
+	Declaration const* m_declaration = nullptr;
+	bool m_willBeWrittenTo = false;
+};
+
+struct SimpleCounterForLoopChecker: public PostTypeChecker::Checker
+{
+	SimpleCounterForLoopChecker(ErrorReporter& _errorReporter): Checker(_errorReporter) {}
+	bool visit(ForStatement const& _forStatement) override
+	{
+		_forStatement.annotation().isSimpleCounterLoop = isSimpleCounterLoop(_forStatement);
+		return true;
+	}
+	bool isSimpleCounterLoop(ForStatement const& _forStatement) const
+	{
+		auto const* cond = dynamic_cast<BinaryOperation const*>(_forStatement.condition());
+		if (!cond || cond->getOperator() != Token::LessThan || cond->userDefinedFunctionType())
+			return false;
+		if (!_forStatement.loopExpression())
+			return false;
+
+		auto const* post = dynamic_cast<UnaryOperation const*>(&_forStatement.loopExpression()->expression());
+		// This matches both operators ++i and i++
+		if (!post || post->getOperator() != Token::Inc || post->userDefinedFunctionType())
+			return false;
+
+		auto const* lhsIdentifier = dynamic_cast<Identifier const*>(&cond->leftExpression());
+		auto const* lhsType = dynamic_cast<IntegerType const*>(cond->leftExpression().annotation().type);
+		auto const* commonType = dynamic_cast<IntegerType const*>(cond->annotation().commonType);
+
+		if (lhsIdentifier && lhsType && commonType && *lhsType == *commonType)
+		{
+			LValueChecker lValueChecker{*lhsIdentifier};
+			_forStatement.body().accept(lValueChecker);
+			if (!lValueChecker.willBeWrittenTo())
+				return true;
```

Comment:
Is checking `willBeWrittenTo` enough to account for all possibilities where an integer variable could be modified? I think that inline assembly may be a case we're missing here. I don't think Yul assignments get Solidity `Expression` nodes with that annotation which would mean we're not checking them. Perhaps we should skip loops containing inline assembly.

In any case, it would be nice to have a test for this that includes inline assembly to see for sure what the behavior is.

Reply:
That's probably right, I don't think it currently accounts for that.

It should be easy to add a visit for ``InlineAssembly`` in the ``LValueChecker`` and just walk through the ``externalReferences`` in its annotation, checking if it refers to the declaration in question and consider any reference there as a write. That'll also include read-only access from assembly, but that's fine for now, we can still refine later.